walk-in-interview/src/main/java/com/google/job/servlets/JobServlet.java

```diff
@@ -0,0 +1,162 @@
+    }
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) {
+        try {
+            // Gets job post from the form
+            Job job = parseJobPost(request);
+
+            // Stores job post into the database
+            storeJobPost(job);
+
+            // Sends the success status code in the response
+            response.setStatus(HttpServletResponse.SC_OK);
+        } catch (IOException | IllegalArgumentException | ServletException e) {
+            // Sends the fail status code in the response
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+        }
+    }
+
+    /** Handles the PATCH request from client. */
+    public void doPatch(HttpServletRequest request, HttpServletResponse response) {
+        try {
+            // Gets the target job post id from the form
+            String jobId = getJobId(request);
+
+            // Gets job post from the form
+            Job updatedJob = parseUpdatedJobPost(request);
+
+            // Stores job post into the database
+            updateJobPost(jobId, updatedJob);
+
+            // Sends the success status code in the response
+            response.setStatus(HttpServletResponse.SC_OK);
+        } catch (IOException | IllegalArgumentException | ServletException e) {
+            // Sends the fail status code in the response
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+        }
+    }
+
+    /** Parses into Job object from json received from client. */
+    private Job parseJobPost(HttpServletRequest request) throws IOException, IllegalArgumentException {
+        // Parses job object from the POST request
+        BufferedReader bufferedReader = request.getReader();
+        String jobPostJsonStr = bufferedReader.lines().collect(Collectors.joining(System.lineSeparator())).trim();
+
+        if (StringUtils.isBlank(jobPostJsonStr)) {
+            throw new IllegalArgumentException("Json for Job object is Empty");
+        }
+
+        Job rawJob = ServletUtils.parseFromJsonUsingGson(jobPostJsonStr, Job.class);
+
+        // Sets the status to be ACTIVE and validates the attributes via build().
+        Job job = rawJob.toBuilder().setJobStatus(JobStatus.ACTIVE).build();
+
+        return job;
+    }
+
+    /** Stores the job post into the database. */
+    private void storeJobPost(Job job) throws ServletException, IOException {
+        try {
+            // Synchronizes and blocks the operation.
+            this.jobsDatabase.addJob(job).get();
+        } catch (InterruptedException e) {
+            throw new ServletException(e);
+        } catch (ExecutionException e) {
+            String errorMessage = "Error occur: " + e.getCause();
+            System.err.println(errorMessage);
+            throw new IOException(errorMessage);
+        }
+    }
+
+    /** Gets the job id from the client. */
+    private String getJobId(HttpServletRequest request) throws ServletException, IOException {
```

Comment:
This method does 2 things. Get the job id from the request, which it's supposed to do based on its name.
And then verify if the id is in the database. Maybe that second logic could be in a different method? E.g verifyUserCanUpdateJob(user, jobId)?